---
title: "Cannabis Expression Atlas - Methods"
author: "Kevelin Barbosa-Xavier, Francisnei Pedrosa-Silva, Fabricio Almeida-Silva, Thiago M. Venancio"
date: "2024-08-22"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = FALSE, warning = FALSE, message = FALSE)
```

## **Exploratory data analysis**

- load important packages
```{r here}
set.seed(123) # for reproducibility
# Load required packages
library(here)
library(tidyverse)
library(patchwork)
library(maps)
library(ggrepel)
library(dplyr)
library(stringr)
library(magrittr)
```

- choose the work directoy
```{r echo = F}
data <- "/media/winterfell/kevelin/doutorado/analises_JLmY_reference/CSEA_LATEST_08_2024/kevelin/"
```
```{r}
data <- data <- "/your/path/here/"
```

### Number of samples per tissue

- load data
```{r}
load("data/sample_metadata_new.rda")
```

```{r}
samples_per_Tissue <- sample_metadata_new %>%
  group_by(Tissue) %>%
  summarise(n = n()) %>%
  arrange(-n) %>%
  mutate(
    Tissue = str_to_title(Tissue),
    Tissue = factor(Tissue, levels = unique(Tissue))
  )

# Plot data
p_samples_per_Tissue <- ggplot(
    samples_per_Tissue, aes(x = n, y = Tissue)
) +
    geom_col(fill = "#053D38") +
    geom_text(aes(label = n), hjust = -0.4, size = 4) +  
    theme_minimal() +
    theme(
        text = element_text(size = 14, colour = "black"),
        axis.text = element_text(size = 14, colour = "black"),
        axis.title = element_text(size = 14),
        plot.title = element_text(size = 16, face = "bold") 
    ) +
    labs(
        x = "Samples", y = "",
        title = "Number of samples per tissue"
    ) +
    scale_x_continuous(limits = c(0, 161))

p_samples_per_Tissue

# Save plot to file
save(
    p_samples_per_Tissue, compress = "xz",
    file = here(data, "data/p_samples_per_Tissue.rda")
)
```

### Number of samples per chemotype
```{r}
samples_per_chemotype <- sample_metadata_new %>%
  group_by(Chemotype) %>%
  summarise(n = n()) %>%
  arrange(-n) %>%
  mutate(
    Chemotype = factor(Chemotype, levels = unique(Chemotype))
  )

# Plot data
p_samples_per_chemotype <- ggplot(
    samples_per_chemotype, aes(x = n, y = Chemotype)
) +
    geom_bar(stat = "identity", fill = "#053D38") +
    geom_text(aes(label = n), hjust = -0.4, size = 4) +  
    theme_minimal() +
    theme(
        text = element_text(size = 14, colour = "black"),
        axis.text = element_text(size = 14, colour = "black"),
        axis.title = element_text(size = 14),
        plot.title = element_text(size = 16, face = "bold") 
    ) +
    labs(
        x = "Samples", y = "",
        title = "Number of Samples per chemotype"
    ) +
    scale_x_continuous(limits = c(0, 271))

p_samples_per_chemotype

# Save plot to file
save(
    p_samples_per_chemotype, compress = "xz",
    file = here(data, "data/p_samples_per_chemotype.rda")
)
```

- p_sample_per_tissue_chemotype
```{r}
p_sample_per_tissue_chemotype <- cowplot::plot_grid(
 p_samples_per_Tissue, p_samples_per_chemotype,
  nrow = 2)
```

### Number of samples per cultivar
```{r}
samples_per_cultivar <- sample_metadata_new %>%
  group_by(Cultivar) %>%
  summarise(n = n()) %>%
  arrange(-n) %>%
  mutate(
    Cultivar = str_to_title(Cultivar),
    Cultivar = factor(Cultivar, levels = unique(Cultivar))
  )

p_samples_per_Cultivar <- ggplot(
    samples_per_cultivar, aes(x = n, y = Cultivar)
) +
    geom_bar(stat = "identity", fill = "#053D38") +
    geom_text(aes(label = n), hjust = -1, size = 4) + 
    theme_minimal() +
    theme(
        text = element_text(size = 14, color = "black"),
        axis.text = element_text(size = 14, color = "black"),
        axis.title = element_text(size = 14, color = "black"),
        plot.title = element_text(size = 16, face = "bold", color = "black")) +
    labs(
        x = "Samples", y = "",
        title = "Number of Samples per Cultivar"
    ) +
    scale_x_continuous(limits = c(0, 50))

p_samples_per_Cultivar

# Save plot to file
save(
    p_samples_per_Cultivar, compress = "xz",
    file = here(data, "data/p_samples_per_Cultivar.rda")
)
```

### Number of samples per sequencing technology and library layout
```{r sequencing-tech-and-layout}
# Get a data frame of counts for sequencing technology
seqtech_count <- sample_metadata_new %>%
   mutate(
    Instrument = str_replace_all(
      Instrument, c(
        "^HiSeq X Ten$" = "Illumina HiSeq X Ten",
        "^Illumina HiSeq X$" = "Illumina HiSeq X Ten"
      )
    )
  ) %>%
  group_by(Instrument) %>%
  summarise(n = n()) %>%
  arrange(-n) %>%
  mutate(
    Instrument = factor(Instrument, levels = unique(Instrument)),
    Group = "Sequencing instrument"
  ) %>%
  rename("Variable" = Instrument)

# Get a data frame of counts for library layout
layout_count <- sample_metadata_new %>%
  group_by(Layout) %>%
  summarise(n = n()) %>%
  arrange(-n) %>%
  mutate(
    Layout = factor(Layout, levels = unique(Layout)),
    Group = "Layout"
  ) |>
    rename("Variable" = Layout)

# Plot data
p_seqtech_layout_count <- bind_rows(
    seqtech_count, layout_count
) |>
    mutate(prop = n / nrow(sample_metadata_new)) |>
    ggplot(aes(x = prop, y = Variable)) +
    geom_bar(aes(fill = Group), stat = "identity", show.legend = FALSE) +
    ggsci::scale_fill_jama() +
    facet_grid(Group ~ ., scales = "free", space = "free") +
    geom_text(aes(label = n), hjust = -0.2) +
    theme_bw() +
    scale_x_continuous(limits = c(0, 1)) +
    labs(
        title = "Summary statistics on library layout and sequencing instrument",
        x = "Relative frequency", y = ""
    )+
  theme(
        text = element_text(size = 14, color = "black"),
        axis.text = element_text(size = 14, color = "black"),
        axis.title = element_text(size = 14, color = "black"),
        plot.title = element_text(size = 16, face = "bold", color = "black"))
p_seqtech_layout_count

# Save plot to file
save(
    p_seqtech_layout_count, compress = "xz",
    file = here(data, "data/p_seqtech_layout_count.rda")
)
```

### Number of samples over time
```{r}
p_nsamples_over_time_cumsum <- sample_metadata_new |>
    dplyr::select(Date) |>
    dplyr::mutate(
        Date = as.Date(Date),
        N = 1
    ) |>
    dplyr::arrange(Date) |>
    dplyr::mutate(cumulative_sum = cumsum(N)) |>
    ggplot(aes(x = Date, y = cumulative_sum, group = 1)) +
    geom_line(colour = "#053D38", size = 2) +
    scale_x_date(date_breaks = "2 year", date_labels = "%Y") + 
    theme_minimal() +
    labs(
        title = "Cumulative number of samples over time",
        x = "", y = "Samples") +
  theme(
        text = element_text(size = 14, color = "black"),
        axis.text = element_text(size = 14, color = "black"),
        axis.title = element_text(size = 14, color = "black"),
        plot.title = element_text(size = 16, face = "bold", color = "black"))


p_nsamples_over_time_cumsum

# Save plot object
save(
    p_nsamples_over_time_cumsum, compress = "xz",
    file = here(data, "data/p_nsamples_over_time_cumsum.rda")
)
```

### Number of samples for each tissue over time
Here we consider just tissues with at least 10 samples
```{r}
p_tissue_timeseries <- sample_metadata_new %>%
    dplyr::select(Tissue, Date) %>%
    dplyr::mutate(
        Tissue = str_to_title(Tissue),
        Date = as.Date(Date),
        N = 1
    ) %>%
    dplyr::mutate(
        Tissue = factor(
            Tissue, levels = dplyr::group_by(sample_metadata_new, Tissue) %>% 
              summarise(n = n()) %>%  
              arrange(-n) %>%
                dplyr::filter(n >= 10) %>% 
                dplyr::mutate(Tissue = str_to_title(Tissue)) %>%
                pull(Tissue)
        )
    ) %>%
    dplyr::arrange(Date) %>%
    dplyr::group_by(Tissue) %>%
    dplyr::mutate(cumulative_sum = cumsum(N)) %>%
  filter(!is.na(Tissue)) %>%
    dplyr::ungroup() %>%
    ggplot(aes(x = Date, y = cumulative_sum, group = Tissue)) +
    geom_line(aes(colour = Tissue)) +
    scale_x_date(date_breaks = "2 year", date_labels = "%Y") + 
    theme_minimal() +
    ggsci::scale_color_d3("category20") +
  theme(
        text = element_text(size = 14, color = "black"),
        axis.text = element_text(size = 14, color = "black"),
        axis.title = element_text(size = 14, color = "black"),
        plot.title = element_text(size = 16, face = "bold", color = "black")) +
    labs(
        title = "Cumulative number of samples for each tissue",
        x = "", y = "Samples"
    ) +
    theme(legend.position = "right") +
    guides(color = guide_legend(ncow = 1))

# Save plot object
save(
    p_tissue_timeseries, compress = "xz",
    file = here(data, "data/p_tissue_timeseries.rda")
)
```

### Number of samples of each chemotype over time
```{r}
p_Chemotype_timeseries <- sample_metadata_new %>%
    dplyr::select(Chemotype, Date) %>%
    dplyr::mutate(
        Chemotype = Chemotype,
        Date = as.Date(Date),
        N = 1
    ) %>%
    dplyr::mutate(
        Chemotype = factor(
            Chemotype, levels = dplyr::group_by(sample_metadata_new, Chemotype) %>% 
              summarise(n = n()) %>%  
              arrange(-n) %>%
                dplyr::filter(n >= 10) %>% 
                dplyr::mutate(Chemotype = Chemotype) %>%
                pull(Chemotype)
        )
    ) %>%
    dplyr::arrange(Date) %>%
    dplyr::group_by(Chemotype) %>%
  filter(!is.na(Chemotype)) %>%
    dplyr::mutate(cumulative_sum = cumsum(N)) %>%
    dplyr::ungroup() %>%
    ggplot(aes(x = Date, y = cumulative_sum, group = Chemotype)) +
    geom_line(aes(colour = Chemotype)) +
    scale_x_date(date_breaks = "2 year", date_labels = "%Y") + 
    theme_minimal() +
    ggsci::scale_color_d3("category20") +
  theme(
        text = element_text(size = 14, color = "black"),
        axis.text = element_text(size = 14, color = "black"),
        axis.title = element_text(size = 14, color = "black"),
        plot.title = element_text(size = 16, face = "bold", color = "black")) +
    labs(
        title = "Cumulative number of samples for each chemotype",
        x = "", y = "Samples"
    ) +
    theme(legend.position = "right") +
    guides(color = guide_legend(ncol = 1))

# Save plot object
save(
    p_Chemotype_timeseries, compress = "xz",
    file = here(data, "data/p_Chemotype_timeseries.rda")
)
```

```{r}
p_time_series_tissue_and_chemotype <- cowplot::plot_grid(
  p_tissue_timeseries, p_Chemotype_timeseries,
  nrow = 2)

p_time_series_tissue_and_chemotype

# Save plot object
save(
    p_time_series_tissue_and_chemotype, compress = "xz",
    file = here(data, "data/p_time_series_tissue_and_chemotype.rda")
)
```

### Number of samples per country

Geographic information will be obtainded from the 
file `SraRunTable_08_2024_cannabisa.txt`, which was downloaded from SRA Run
Selector. In details, we searched "cannabis sativa[ORGN] AND RNA-seq[STRA]"
on NCBI's SRA, clicked on *Send results to Run selector*, and downloaded 
the metadata for all samples.
```{r}
#load data
samples_per_country <- read_csv(
    here(data, "data/SraRunTable_08_2024_cannabisa.txt")
) 
```

We need to prepare the data because some informations for country need o be cureted by hand.
```{r}
sample_metadata_new$BioSample <- row.names(sample_metadata_new)
samples_per_country <- inner_join(samples_per_country, sample_metadata_new, by = "BioSample") |>
  distinct(BioSample, .keep_all = TRUE) |>
  dplyr::select(BioSample, BioProject.x, `Center Name`, `geo_loc_name_country`, `geo_loc_name`) |>
  arrange(`geo_loc_name_country`) |>
  mutate(`geo_loc_name_country` = ifelse(is.na(`geo_loc_name_country`), `Center Name`, `geo_loc_name_country`)) |>
  mutate(`geo_loc_name_country` = ifelse(str_detect(geo_loc_name_country, "uncalculated"), `Center Name`, `geo_loc_name_country`)) |>
  mutate(`geo_loc_name_country` = ifelse(str_detect(geo_loc_name_country, "GEO"), `BioProject.x`, `geo_loc_name_country`)) |>
  mutate(
        Country = str_replace_all(
            `geo_loc_name_country`, c(
  "MARC_BEHR'S SHARED SUBMISSIONS" = "Luxembourg",
  "CHINA ACADEMY OF CHINESE MEDICAL SCIENCES" = "China",
  "UNIVERSITY OF TORONTO" = "Canada",
  "MICHIGAN STATE UNIVERSITY" = "USA",
  "DEPARTMENT OF BIOLOGICAL SCIENCES" = "China",
  "PRJNA245084" = "China",
  "PRJNA756306" = "USA",
  "PRJNA435671" = "Luxembourg",
  "LA TROBE UNIVERSITY" = "Australia",
  "Genome Quebec" = "Quebec",
  "DEPARTMENT OF BIO-HEALTH CONVERGENCE, KANGWON NATIONAL UNIVSERSITY" = "Republic of Korea")))
```

```{r}
final_samples_per_country <- samples_per_country |>
    dplyr::count(Country) |>
    arrange(-n)

save(
    final_samples_per_country, compress = "xz",
    file = here(data, "data/final_samples_per_country.rda")
)
```

```{r}
# Plot bar plot of number of samples per country
final_samples_per_country$n <- as.numeric(final_samples_per_country$n)
p_barplot_samples_per_country <- final_samples_per_country |>
    mutate(Country = factor(Country, levels = rev(Country))) |>
    ggplot(aes(x = n, y = Country)) +
    geom_col(fill = "#053D38") +
    geom_text(aes(label = n), hjust = -0.2) +
    theme_minimal() +
    labs(
        title = "Number of samples per country",
        x = "Samples", y = ""
    ) + 
    xlim(0, 110)+
  theme(
        text = element_text(size = 14, color = "black"),
        axis.text = element_text(size = 14, color = "black"),
        axis.title = element_text(size = 14, color = "black"),
        plot.title = element_text(size = 16, face = "bold", color = "black"))

p_barplot_samples_per_country

save(
    p_barplot_samples_per_country, compress = "xz",
    file = here(data, "data/p_barplot_samples_per_country.rda")
)
```

### Distribution of Mean Read Lengths per Sample

```{r}
load("data/fastp_stats.rda")
load("data/mapping_rate.rda")
```

```{r}
# Calculate density
mean_length_density <- density(fastp_stats$before_meanlength)

# Create a data frame with the density results
density_data <- data.frame(x = mean_length_density$x, y = mean_length_density$y)

p_mean_length_density <- ggplot(density_data, aes(x = x, y = y)) +
  geom_polygon(fill = "lightblue", color = "black") +
  geom_vline(xintercept = 40, linetype = "dashed", color = "red") +
  labs(title = "Distribution of mean read lengths per Sample",
       y = "Density",
       x = "Mean Read Lengths") +
  theme_minimal() +
  theme(
        text = element_text(size = 14, color = "black"),
        axis.text = element_text(size = 14, color = "black"),
        axis.title = element_text(size = 14, color = "black"),
        plot.title = element_text(size = 16, face = "bold", color = "black"))

p_mean_length_density

save(
  p_mean_length_density,
  file = here::here(data, "data/p_mean_length_density.rda")
)
```

```{r}
# Calculate density
q20_density <- density(fastp_stats$before_q20rate)

# Create a data frame with the density results
q20_density_data <- data_frame(x = q20_density$x, y = as.numeric(q20_density$y))

p_q20_density <- ggplot(q20_density_data, aes(x = x, y = y)) +
  geom_polygon(fill = "lightgreen", color = "black") +
  geom_vline(xintercept = 0.8, linetype = "dashed", color = "red") +
  labs(title = "Distribution of Q20 rates per sample",
       y = "Density",
       x = "Q20 Rates") +
  theme_minimal()+
  theme(
        text = element_text(size = 14, color = "black"),
        axis.text = element_text(size = 14, color = "black"),
        axis.title = element_text(size = 14, color = "black"),
        plot.title = element_text(size = 16, face = "bold", color = "black"))

p_q20_density

save(
  p_q20_density,
  file = here::here(data, "data/p_q20_density.rda")
)
```

```{r}
# remove NA
clean_mapping_rate <- na.omit(mapping_rate$Mapping_rate)

# Calculate density
mapping_rate_density <- density(clean_mapping_rate)

# Create a data frame with the density results
mapping_rate_density_data <- data.frame(x = mapping_rate_density$x, y = mapping_rate_density$y)

p_mapping_rate_density <- ggplot(mapping_rate_density_data, aes(x = x, y = y)) +
  geom_polygon(fill = "lightyellow", color = "black") +
  geom_vline(xintercept = 50, linetype = "dashed", color = "red") +
  labs(title = "Distribution of mapping rates per sample",
       y = "Density",
       x = "Mapping Rates") +
  theme_minimal()+
  theme(
        text = element_text(size = 14, color = "black"),
        axis.text = element_text(size = 14, color = "black"),
        axis.title = element_text(size = 14, color = "black"),
        plot.title = element_text(size = 16, face = "bold", color = "black"))

p_mapping_rate_density

save(
  p_mapping_rate_density,
  file = here::here(data, "data/p_mapping_rate_density.rda")
)
```

```{r}
# Calculate density
nreads_density <- density(fastp_stats$after_nreads / 1000000)

nreads_median <- (median(fastp_stats$after_nreads) / 1000000)
nreads_mean <- (mean(fastp_stats$after_nreads ) / 1000000)

# Create a data frame with the density results
nreads_density_data <- data.frame(x = nreads_density$x, y = nreads_density$y)

p_nreads_density <- ggplot(nreads_density_data, aes(x = x, y = y)) +
  geom_polygon(fill = "lightpink", color = "black") +
  labs(title = "Distribution of Number of Reads per Sample",
       y = "Density",
       x = "Number of Reads (Millions)") +
  theme_minimal()+
 theme(
        text = element_text(size = 14, color = "black"),
        axis.text = element_text(size = 14, color = "black"),
        axis.title = element_text(size = 14, color = "black"),
        plot.title = element_text(size = 16, face = "bold", color = "black"))
```

- Fig_summary_quality_statistics
```{r}
#plot
library(cowplot)
p_summary_quality_statistics <- cowplot::plot_grid(
  p_mean_length_density, p_q20_density, p_mapping_rate_density, p_nreads_density,
  nrow = 2
)

p_summary_quality_statistics

save(
  p_summary_quality_statistics,
  file = here::here(data, "data/p_summary_quality_statistics.rda")
)
```

